kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/evm-int-simplification.k

```diff
@@ -57,31 +50,91 @@ module EVM-INT-SIMPLIFICATION-COMMON
+    [simplification(40), preserves-definedness]
+
+  // Reduction: `#range` can be ignored under certain conditions
+  rule [asWord-range]:
+    #asWord ( #range ( B:Bytes, S, W ) ) => #asWord ( B )
     requires 0 <=Int S andBool 0 <=Int W
+     andBool lengthBytes(B) <=Int 32
      andBool lengthBytes(B) ==Int S +Int W
      andBool #asWord ( B ) <Int 2 ^Int ( 8 *Int W )
-    [simplification, concrete(X, S, W), preserves-definedness]
+    [simplification, concrete(S, W), preserves-definedness]
+
+  // Reduction: `#asWord` inverts `#buf` under certain conditions
+  rule [asWord-buf-inversion]:
+    #asWord ( #buf ( WB:Int, X:Int ) ) => X
+    requires 0 <=Int WB andBool 0 <=Int X andBool X <Int minInt ( 2 ^Int (8 *Int WB), pow256 )
+    [simplification, concrete(WB), preserves-definedness]
+
+  // Reduction: bitwise right shift in terms of `#range`
+  rule [asWord-shr]:
+    #asWord( BA ) >>Int N => #asWord ( #range ( BA, 0, lengthBytes( BA ) -Int ( N /Int 8 ) ) )
+    requires 0 <=Int N andBool N modInt 8 ==Int 0
+    [simplification, concrete(N), preserves-definedness]
+
+  // Reduction: division in terms of `#range`
+  rule [asWord-div]:
+    #asWord( BA ) /Int N => #asWord ( #range ( BA, 0, lengthBytes( BA ) -Int ( log2Int( N ) /Int 8 ) ) )
+    requires 0 <Int N andBool N ==Int 2 ^Int log2Int ( N ) andBool log2Int( N ) modInt 8 ==Int 0
+    [simplification, preserves-definedness]
+
+  // Comparison: `#asWord` is always in the range `[0, pow256)`
+  rule [asWord-lb]: 0 <=Int #asWord( _ )     => true [simplification, smt-lemma]
+  rule [asWord-ub]: #asWord( _ ) <Int pow256 => true [simplification, smt-lemma]
+
+  // Comparison: `#asWord(B)` is certainly less than something that does not fit into `lengthBytes(B)` bytes
+  rule [asWord-lt]:  #asWord ( B )  <Int X:Int => true  requires X >=Int   2 ^Int (8 *Int minInt(32, lengthBytes(B)))               [simplification, preserves-definedness]
+  // Comparison: `#asWord(B)` is certainly less than or equal to something that is not less than the max number fitting into `lengthBytes(B)` bytes
+  rule [asWord-le]:  #asWord ( B ) <=Int X:Int => true  requires X >=Int ( 2 ^Int (8 *Int minInt(32, lengthBytes(B))) -Int 1 )      [simplification, preserves-definedness]
+
+  // Comparison: `#asWord` of `+Bytes` when lower bytes match, with `<Int`
+  rule [asWord-concat-lt]:
+    #asWord ( BA1 +Bytes BA2 ) <Int X:Int => #asWord ( BA1 ) <Int X /Int ( 2 ^Int ( 8 *Int lengthBytes ( BA2 ) ) )
+    requires #asWord ( BA2 ) ==Int X modInt ( 2 ^Int ( 8 *Int lengthBytes ( BA2 ) ) )
+    [simplification, concrete(BA2, X), preserves-definedness]
+
+  // Comparison: `#asWord` of `+Bytes` when lower bytes match, with `<=Int`
+  rule [asWord-concat-le]:
+    #asWord ( BA1 +Bytes BA2 ) <=Int X:Int => #asWord ( BA1 ) <=Int X /Int ( 2 ^Int ( 8 *Int lengthBytes ( BA2 ) ) )
+    requires #asWord ( BA2 ) ==Int X modInt ( 2 ^Int ( 8 *Int lengthBytes ( BA2 ) ) )
+    [simplification, concrete(BA2, X), preserves-definedness]
```

Comment:
I think these both suffer similar counterexamples:

- `BA1 == #buf(1, 8)`
- `BA2 == #buf(64, maxUInt512)`
- `X == pow514 +Int maxUInt512`

Reduction 1:

```
   #asWord(BA1 +Bytes BA2) <Int X
=> maxUInt256 <Int pow514 +Int maxUInt512
=> true
```

Reduction 2:

```
   #asWord(BA1 +Bytes BA2) <Int X
=> #asWord(BA1) <Int X /Int (2 ^Int (8 *Int lengthBytes(BA2)))
=> 8 <Int (pow514 +Int maxUInt512) /Int pow512
=> 8 <Int pow2
=> 8 <Int 4
=> false
```

I think what's needed here are restrictions on the length of `BA1 +Bytes BA2`?

Reply:
This can be added as a test.

Reply:
I think that we need to put higher priority on the simplification that cuts a buffer inside asWord to be of max length 32.

Reply:
It is of higher priority already. I will see what the tests do.

Reply:
I have added two very general tests that show that the part of the buffer overflowing 32 bytes will be chopped off correctly before the simplification is applied.

Reply:
Same as above, we cannot rely on priority of `simplification` rules, it's a suggestion for the backend, not a requirement.

Reply:
Corrected.